external/urlfrontier/src/test/java/com/digitalpebble/stormcrawler/urlfrontier/StatusUpdaterBoltTest.java

```diff
@@ -0,0 +1,127 @@
+    private StatusUpdaterBolt bolt;
+    private TestOutputCollector output;
+    private URLFrontierContainer urlFrontierContainer;
+
+    private static final String persistedKey = "somePersistedKey";
+    private static final String notPersistedKey = "someNotPersistedKey";
+
+    private static ExecutorService executorService;
+
+    @Rule public Timeout globalTimeout = Timeout.seconds(60);
+
+    @BeforeClass
+    public static void beforeClass() {
+        executorService = Executors.newFixedThreadPool(2);
+    }
+
+    @AfterClass
+    public static void afterClass() {
+        executorService.shutdown();
+        executorService = null;
+    }
+
+    @Before
+    public void before() {
+
+        String version = System.getProperty("urlfrontier-version");
+        if (version == null) version = "2.2";
+
+        urlFrontierContainer = new URLFrontierContainer("crawlercommons/url-frontier:" + version);
+        urlFrontierContainer.start();
+
+        bolt = new StatusUpdaterBolt();
+
+        var connection = urlFrontierContainer.getFrontierConnection();
+
+        final var config = new HashMap<String, Object>();
+        config.put(
+                "urlbuffer.class",
+                "com.digitalpebble.stormcrawler.persistence.urlbuffer.SimpleURLBuffer");
+        config.put(Constants.URLFRONTIER_HOST_KEY, connection.getHost());
+        config.put(Constants.URLFRONTIER_PORT_KEY, connection.getPort());
+        config.put(
+                "scheduler.class", "com.digitalpebble.stormcrawler.persistence.DefaultScheduler");
+        config.put("status.updater.cache.spec", "maximumSize=10000,expireAfterAccess=1h");
+        config.put("metadata.persist", persistedKey);
+
+        output = new TestOutputCollector();
+        bolt.prepare(config, TestUtil.getMockedTopologyContext(), new OutputCollector(output));
+    }
+
+    @After
+    public void after() {
+        bolt.cleanup();
+        urlFrontierContainer.close();
+        output = null;
+    }
+
+    @SuppressWarnings("SameParameterValue")
+    private Future<Integer> store(String url, Status status, Metadata metadata) {
+        Tuple tuple = mock(Tuple.class);
+        when(tuple.getValueByField("status")).thenReturn(status);
+        when(tuple.getStringByField("url")).thenReturn(url);
+        when(tuple.getValueByField("metadata")).thenReturn(metadata);
+        bolt.execute(tuple);
+
+        return executorService.submit(
+                () -> {
+                    var outputSize = output.getAckedTuples().size();
+                    while (outputSize == 0) {
+                        Thread.sleep(100);
+                        outputSize = output.getAckedTuples().size();
+                    }
+                    return outputSize;
+                });
+    }
+
+    @Test
+    public void canAckASimpleTuple()
+            throws ExecutionException, InterruptedException, TimeoutException {
+
+        Configurator.setLevel(StatusUpdaterBolt.class, Level.ALL);
```

Comment:
log4j - please remove alongside any other use of it and rely on slf4j instead

Reply:
slf4j does not give the option to set a global logging level like log4j.
Should I drop it?

Reply:
what about something like this (not tested)

 ```
       LoggerContext loggerContext = (LoggerContext) LoggerFactory.getILoggerFactory();
        loggerContext.getLogger("com.digitalpebble.stormcrawler.urlfrontier").setLevel(Level.TRACE);
        
```